src/main/flight/rpm_filter.c

```diff
@@ -45,200 +45,155 @@
+    if (!config->rpm_filter_harmonics) {
+        return;
+    }
 
-    const float loopIterationsPerUpdate = MIN_UPDATE_T / (pidLooptime * 1e-6f);
-    numberFilters = getMotorCount() * (filters[0].harmonics + filters[1].harmonics);
-    const float filtersPerLoopIteration = numberFilters / loopIterationsPerUpdate;
-    filterUpdatesPerIteration = lrintf(filtersPerLoopIteration + 0.49f);
-}
+    // if we get to this point, enable and init RPM filtering
+    rpmFilter.numHarmonics = config->rpm_filter_harmonics;
+    rpmFilter.minHz = config->rpm_filter_min_hz;
+    rpmFilter.maxHz = 0.48f * 1e6f / looptimeUs; // don't go quite to nyquist to avoid oscillations
+    rpmFilter.fadeRangeHz = config->rpm_filter_fade_range_hz;
+    rpmFilter.q = config->rpm_filter_q / 100.0f;
+    rpmFilter.looptimeUs = looptimeUs;
 
-static float applyFilter(rpmNotchFilter_t *filter, const int axis, float value)
-{
-    if (filter == NULL) {
-        return value;
-    }
-    for (int motor = 0; motor < getMotorCount(); motor++) {
-        for (int i = 0; i < filter->harmonics; i++) {
-            value = biquadFilterApplyDF1Weighted(&filter->notch[axis][motor][i], value);
+    for (int axis = 0; axis < XYZ_AXIS_COUNT; axis++) {
+        for (int motor = 0; motor < getMotorCount(); motor++) {
+            for (int i = 0; i < rpmFilter.numHarmonics; i++) {
+                biquadFilterInit(&rpmFilter.notch[axis][motor][i], rpmFilter.minHz * i, rpmFilter.looptimeUs, rpmFilter.q, FILTER_NOTCH, 0.0f);
+            }
         }
     }
-    return value;
-}
 
-float rpmFilterGyro(const int axis, float value)
-{
-    return applyFilter(gyroFilter, axis, value);
+    erpmToHz = ERPM_PER_LSB / SECONDS_PER_MINUTE  / (motorConfig()->motorPoleCount / 2.0f);
+
+    const float loopIterationsPerUpdate = RPM_FILTER_DURATION_S / (looptimeUs * 1e-6f);
+    const float numNotchesPerAxis = getMotorCount() * rpmFilter.numHarmonics;
+    notchUpdatesPerIteration = ceilf(numNotchesPerAxis / loopIterationsPerUpdate); // round to ceiling
 }
 
 FAST_CODE_NOINLINE void rpmFilterUpdate(void)
 {
+    if (!motorConfig()->dev.useDshotTelemetry) {
+        return;
+    }
+
+    // update motor RPM data
+    minMotorFrequencyHz = FLT_MAX;
     for (int motor = 0; motor < getMotorCount(); motor++) {
-        filteredMotorErpm[motor] = pt1FilterApply(&rpmFilters[motor], getDshotTelemetry(motor));
+        motorFrequencyHz[motor] = pt1FilterApply(&motorFreqLpf[motor], erpmToHz * getDshotTelemetry(motor));
+        minMotorFrequencyHz = MIN(minMotorFrequencyHz, motorFrequencyHz[motor]);
         if (motor < 4) {
-            DEBUG_SET(DEBUG_RPM_FILTER, motor, motorFrequency[motor]);
+            DEBUG_SET(DEBUG_RPM_FILTER, motor, motorFrequencyHz[motor]);
         }
-        motorFrequency[motor] = erpmToHz * filteredMotorErpm[motor];
     }
 
-    if (gyroFilter == NULL) {
-        minMotorFrequency = 0.0f;
+    if (!isRpmFilterEnabled()) {
         return;
     }
```

Comment:
Should this be moved before update motor RPM data? (line 131)

Reply:
No, that's intended because of historical reasons. Getting RPM telemetry was (and still is) part of the RPM filter. To make use of RPM data while RPM Filtering is disabled we need this split.

    So first you need to check if bidir Dshot is active
    then you initialize the RPM data lowpass filters
    then you check if RPM Filter is active
    and if so you initialize the RPM Filter

I am planning to refactor the Dshot part out of the RPM Filter in BF 4.5 so this doesn't look weird anymore.